lib/chef/resource/osx_profile.rb

```diff
@@ -42,8 +43,245 @@ class OsxProfile < Chef::Resource
+          current_resource.profile_name(new_resource.profile_name)
+
+          if new_profile_hash
+            new_profile_hash["PayloadUUID"] = config_uuid(new_profile_hash)
+          end
+
+          current_resource.profile(current_profile)
+        end
+
+        def current_profile
+          all_profiles = get_installed_profiles
+
+          if all_profiles && all_profiles.key?("_computerlevel")
+            return all_profiles["_computerlevel"].find do |item|
+              item["ProfileIdentifier"] == new_profile_identifier
+            end
+          end
+          nil
+        end
+
+        def check_resource_semantics!
+          if mac? && node["platform_version"] > 10.15
+            raise "The osx_profile resource is not availble on macOS Bug Sur or above due to the removal of apple support for CLI installation of profiles"
+          end
+
+          if action == :remove
+            if new_profile_identifier
+              unless !new_profile_identifier.nil? && !new_profile_identifier.end_with?(".mobileconfig") && /^\w+(?:(\.| )\w+)+$/.match(new_profile_identifier)
+                raise "when removing using the identifier property, it must match the profile identifier"
+              end
+            else
+              new_profile_name = new_resource.profile_name
+              unless !new_profile_name.end_with?(".mobileconfig") && /^\w+(?:(\.| )\w+)+$/.match(new_profile_name)
+                raise "When removing by resource name, it must match the profile identifier"
+              end
+            end
+          end
+
+          if action == :install
+            if new_profile_hash.is_a?(Hash)
+              unless new_profile_hash.include?("PayloadIdentifier")
+                raise "The specified profile does not seem to be valid"
+              end
+            end
+            if new_profile_hash.is_a?(String)
+              unless new_profile_hash.end_with?(".mobileconfig")
+                raise "#{new_profile_hash}' is not a valid profile"
+              end
+            end
+          end
+        end
+      end
+
+      action :install do
+        unless profile_installed?
+          converge_by("install profile #{new_profile_identifier}") do
+            profile_path = write_profile_to_disk
+            install_profile(profile_path)
+            get_installed_profiles(true)
+          end
+        end
+      end
+
+      action :remove do
+        # Clean up profile after removing it
+        if profile_installed?
+          converge_by("remove profile #{new_profile_identifier}") do
+            remove_profile
+            get_installed_profiles(true)
+          end
+        end
+      end
+
+      action_class do
+        private
+
+        def profile
+          @profile ||= new_resource.profile || new_resource.profile_name
+        end
+
+        def new_profile_hash
+          @new_profile_hash ||= get_profile_hash(profile)
+        end
```

Comment:
Could this be set to `{}` if `get_profile_hash` returns nil, or make it so that function always returns a hash? This gets memoized which means it could be run multiple times if its nil, and there's a lot of checking to make sure this function does not return nil.

Reply:
Changing it to `{}` would mean that the nil checks need to be `empty?` checks and without any domain knowledge of what is going on here, i don't know if `{}` is a separate valid case or not.  I thought about it, but being backwards compatible with the edge case seemed to be better for this refactoring than worrying about speed.